README.md
![Release](https://github.com/mmuller88/influxdb-s3-backup/workflows/Release/badge.svg)
![push-docker](https://github.com/mmuller88/influxdb-s3-backup/workflows/push-docker/badge.svg)

# influxdb-s3-backup

A docker container for backing up your InfluxDB to AWS S3. It works on a Raspberry Pi as well.

# Docker Hub

Released to [Docker Hub](https://hub.docker.com/repository/docker/damadden88/influxdb-s3-backup). Supported platforms: linux/amd64,linux/arm64

# Usage

## Default cron (1am daily)

```shell
docker run \
    -e DATABASE=mydatabase \
    -e DATABASE_HOST=1.2.3.4 \
    -e S3_BUCKET=mybackupbucket \
    -e AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE \
    -e AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY \
    -e AWS_DEFAULT_REGION=us-west-2 \
    damadden88/influxdb-to-s3:latest
```

or if you just start your docker compose deployment

## Run Docker Container

```shell
docker run \
    -e DATABASE=mydatabase \
    -e DATABASE_HOST=1.2.3.4 \
    -e S3_BUCKET=mybackupbucket \
    -e AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE \
    -e AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY \
    -e AWS_DEFAULT_REGION=us-west-2 \
    damadden88/influxdb-to-s3:latest
```

## Docker Compose deployment

### Preparation

You need to set you AWS Credentials before like:

```
export AWS_ACCESS_KEY_ID=AKIAxx
export AWS_SECRET_ACCESS_KEY=7PBRxx
```

And than run:

```
docker-compose up -d --build
```

### Example

```yaml
version: '3.3'

services:
  influxdb:
    image: influxdb:latest
    environment:
      INFLUXDB_DB: mydb
      INFLUXDB_BIND_ADDRESS: 0.0.0.0:8088
  influxdbs3backup:
    build:
      context: .
    environment:
      DATABASE: mydb
      DATABASE_HOST: influxdb
      S3_BUCKET: mybackupbucket
      AWS_ACCESS_KEY_ID: AKIAIOSFODNN7EXAMPLE
      AWS_SECRET_ACCESS_KEY: wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY
      AWS_DEFAULT_REGION: us-west-2
      CRON: '* * * * *'
```

### Run backup

```shell
docker run \
    -e DATABASE=mydatabase \
    -e DATABASE_HOST=1.2.3.4 \
    -e S3_BUCKET=mybackupbucket \
    -e AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE \
    -e AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY \
    -e AWS_DEFAULT_REGION=us-west-2 \
    damadden88/influxdb-to-s3:latest \
    backup
```

### Restore

**Warning** - Restores [cannot be run on a running InfluxDB instance](https://docs.influxdata.com/influxdb/v1.1/administration/backup_and_restore/#restore), which precludes doing a remote restore. This means there is a requirement for the restore container to have local access to the influxdb `meta` and `data` directories, which are probably located in `/var/lib/influxdb` unless you've done something non-standard. You must therefore pass in the influxdb directory as a volume instead of specifying the host.

```shell
docker run \
    -v /path/to/influxdb:/var/lib/influxdb \
    -e DATABASE=mydatabase \
    -e S3_BUCKET=mybackupbucket \
    -e AWS_ACCESS_KEY_ID=AKIAIOSFODNN7EXAMPLE \
    -e AWS_SECRET_ACCESS_KEY=wJalrXUtnFEMI/K7MDENG/bPxRfiCYEXAMPLEKEY \
    -e AWS_DEFAULT_REGION=us-west-2 \
    damadden88/influxdb-to-s3:latest \
    restore
```

# Environment Variables

| Variable                |                     Description                      |                 Example Usage |                  Default |                    Optional? |
| ----------------------- | :--------------------------------------------------: | ----------------------------: | -----------------------: | ---------------------------: |
| `DATABASE`              |                  Database to backup                  |                    `telegraf` |                     None |                           No |
| `S3_BUCKET`             |                    Name of bucket                    |                `mybucketname` |                     None |                           No |
| `S3_KEY_PREFIX`         |            S3 directory to place files in            | `backups` or `backups/sqlite` |                     None |                          Yes |
| `AWS_ACCESS_KEY_ID`     |                    AWS Access key                    |                   `AKIAIO...` |                     None | Yes (if using instance role) |
| `AWS_SECRET_ACCESS_KEY` |                    AWS Secret Key                    |              `wJalrXUtnFE...` |                     None | Yes (if using instance role) |
| `AWS_DEFAULT_REGION`    |                  AWS Default Region                  |                   `us-west-2` |              `us-west-1` |                          Yes |
| `DATABASE_HOST`         |         Hostname or IP of influxdb instance          |                     `1.2.3.4` |              `localhost` |                          Yes |
| `DATABASE_PORT`         |              Port of influxdb instance               |                        `8098` |                   `8088` |                          Yes |
| `DATABASE_META_DIR`     |           Path to local influxdb meta dir            |      `/path/to/influxdb/meta` | `/var/lib/influxdb/meta` |                          Yes |
| `DATABASE_DATA_DIR`     |           Path to local influxdb data dir            |      `/path/to/influxdb/data` | `/var/lib/influxdb/data` |                          Yes |
| `BACKUP_PATH`           | Directory to write the backup (within the container) |          `/myvolume/mybackup` |  `/data/influxdb/backup` |                          Yes |
| `BACKUP_ARCHIVE_PATH`   |  Path to compress the backup (within the container)  |      `/myvolume/mybackup.tgz` |     `${BACKUP_PATH}.tgz` |                          Yes |

# Thanks To:

- Jacob Tomlinson https://github.com/jacobtomlinson/docker-influxdb-to-s3
- And to the amazing [Projen Community](https://github.com/projen/projen)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [0.0.57](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.56...v0.0.57) (2021-01-31)


### Bug Fixes

* adjust REadme ([85bbc98](https://github.com/mmuller88/influxdb-s3-backup/commit/85bbc988d7e62c65ed922088220623cd21fadac4))

### [0.0.56](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.55...v0.0.56) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([ca1c824](https://github.com/mmuller88/influxdb-s3-backup/commit/ca1c824a2573b02719a03518bd61d91e203fc096))
* use releaseworkflow for docker hub release ([998b84a](https://github.com/mmuller88/influxdb-s3-backup/commit/998b84ad61c0eed405dd39ef2e43a5f495e48015))

### [0.0.55](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.54...v0.0.55) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([56bcc92](https://github.com/mmuller88/influxdb-s3-backup/commit/56bcc921f6a36d25246b29ed8b6c1bdeb42ceb91))

### [0.0.54](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.53...v0.0.54) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([69d1cb7](https://github.com/mmuller88/influxdb-s3-backup/commit/69d1cb74326074eefc986f42bf90a6f5f5fd24c8))

### [0.0.53](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.52...v0.0.53) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([c4703bf](https://github.com/mmuller88/influxdb-s3-backup/commit/c4703bf5de596058105a42ad9206edc388e49057))

### [0.0.52](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.51...v0.0.52) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([9a40527](https://github.com/mmuller88/influxdb-s3-backup/commit/9a4052789a53800d1d8123cc14d799a8d8573ea0))

### [0.0.51](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.50...v0.0.51) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([70c808e](https://github.com/mmuller88/influxdb-s3-backup/commit/70c808ece9886ee35e8bf6fc824b62d91b90b01b))
* use releaseworkflow for docker hub release ([15969a0](https://github.com/mmuller88/influxdb-s3-backup/commit/15969a04feef053c2dfbba6a8fcbd68826332d90))

### [0.0.50](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.49...v0.0.50) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([8fb7f35](https://github.com/mmuller88/influxdb-s3-backup/commit/8fb7f3511dc1e515cffc307a8785f1f2ab41f56b))

### [0.0.49](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.48...v0.0.49) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([6414579](https://github.com/mmuller88/influxdb-s3-backup/commit/64145794e517406a45f3a08a251aee11ae907147))

### [0.0.48](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.47...v0.0.48) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([6d35342](https://github.com/mmuller88/influxdb-s3-backup/commit/6d353425791b34118e76982b083f9d5ace49d711))

### [0.0.47](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.46...v0.0.47) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([0a8f756](https://github.com/mmuller88/influxdb-s3-backup/commit/0a8f75662b5598ccbbe1e0ed5fa9936bb088562a))

### [0.0.46](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.45...v0.0.46) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([90af6cc](https://github.com/mmuller88/influxdb-s3-backup/commit/90af6cc00e78b9fe56dbb7221dd45fa566ef7d36))

### [0.0.45](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.44...v0.0.45) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([66c7de0](https://github.com/mmuller88/influxdb-s3-backup/commit/66c7de081f088ea824b7a47686ccadf379e1eded))
* use releaseworkflow for docker hub release ([3eff472](https://github.com/mmuller88/influxdb-s3-backup/commit/3eff47271fbf0f570cf42f0df751d930cc75b3b4))

### [0.0.44](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.43...v0.0.44) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([398fc99](https://github.com/mmuller88/influxdb-s3-backup/commit/398fc99e61f34f8304908c5f631793a04c1b5453))
* use releaseworkflow for docker hub release ([25611b6](https://github.com/mmuller88/influxdb-s3-backup/commit/25611b66d2d3dca198a42fc5bf444a2ae9fa9149))

### [0.0.43](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.42...v0.0.43) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([babb5c6](https://github.com/mmuller88/influxdb-s3-backup/commit/babb5c6f359a1d25c4c775e9208637b117cbc0ef))

### [0.0.42](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.41...v0.0.42) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([07eaadb](https://github.com/mmuller88/influxdb-s3-backup/commit/07eaadbc00d970515406020c1c6de3afb2035ac0))

### [0.0.41](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.40...v0.0.41) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([0bc8140](https://github.com/mmuller88/influxdb-s3-backup/commit/0bc814080f56bbd2085790408789036f52b8b3c7))
* use releaseworkflow for docker hub release ([b51f685](https://github.com/mmuller88/influxdb-s3-backup/commit/b51f6859b8ab1c4f854690b9d493ed91181393cc))

### [0.0.40](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.39...v0.0.40) (2021-01-27)


### Features

* use releaseworkflow for docker hub release ([ea2acb9](https://github.com/mmuller88/influxdb-s3-backup/commit/ea2acb9baff382d89af4c30a4f3b906ed35e7258))

### [0.0.39](https://github.com/mmuller88/influxdb-s3-backup/compare/v0.0.38...v0.0.39) (2021-01-27)


### Features


influxdb-to-s3.sh
#!/bin/bash

set -e

export S3_BUCKET=${S3_BUCKET}
# Check and set missing environment vars
: ${S3_BUCKET:?"S3_BUCKET env variable is required"}
: ${AWS_SECRET_ACCESS_KEY:?"AWS_SECRET_ACCESS_KEY env variable is required"}
: ${AWS_ACCESS_KEY_ID:?"AWS_ACCESS_KEY_ID env variable is required"}
if [[ -z ${S3_KEY_PREFIX} ]]; then
  export S3_KEY_PREFIX=""
else
  if [ "${S3_KEY_PREFIX: -1}" != "/" ]; then
    export S3_KEY_PREFIX="${S3_KEY_PREFIX}/"
  fi
fi
: ${DATABASE:?"DATABASE env variable is required"}
export BACKUP_PATH=${BACKUP_PATH:-/data/influxdb/backup}
export BACKUP_ARCHIVE_PATH=${BACKUP_ARCHIVE_PATH:-${BACKUP_PATH}.tgz}
export DATABASE_HOST=${DATABASE_HOST:-localhost}
export DATABASE_PORT=${DATABASE_PORT:-8088}
export DATABASE_META_DIR=${DATABASE_META_DIR:-/var/lib/influxdb/meta}
export DATABASE_DATA_DIR=${DATABASE_DATA_DIR:-/var/lib/influxdb/data}
# export CRON=${CRON:-"0 1 * * *"}
export CRON=${CRON:-"* * * * *"}
export DATETIME=$(date "+%Y%m%d%H%M%S")

# Add this script to the crontab and start crond
startcron() {
  echo "export S3_BUCKET=$S3_BUCKET" >> $HOME/.profile
  echo "export S3_KEY_PREFIX=$S3_KEY_PREFIX" >> $HOME/.profile
  echo "export DATABASE_HOST=$DATABASE_HOST" >> $HOME/.profile
  echo "export DATABASE=$DATABASE" >> $HOME/.profile
  echo "export DATABASE_PORT=$DATABASE_PORT" >> $HOME/.profile
  echo "export BACKUP_PATH=$BACKUP_PATH" >> $HOME/.profile
  echo "export BACKUP_ARCHIVE_PATH=$BACKUP_ARCHIVE_PATH" >> $HOME/.profile
  echo "export DATETIME=$DATETIME" >> $HOME/.profile
  echo "export AWS_DEFAULT_REGION=$AWS_DEFAULT_REGION" >> $HOME/.profile
  echo "export AWS_ACCESS_KEY_ID=$AWS_ACCESS_KEY_ID" >> $HOME/.profile
  echo "export AWS_SECRET_ACCESS_KEY=$AWS_SECRET_ACCESS_KEY" >> $HOME/.profile
  echo "Starting backup cron job with frequency '$1'"

  echo "$1 . $HOME/.profile; $0 backup >> /var/log/cron.log 2>&1" > /etc/cron.d/influxdbbackup

  cat /etc/cron.d/influxdbbackup

  # Apply cron job
  crontab /etc/cron.d/influxdbbackup

  # Create the log file to be able to run tail
  touch /var/log/cron.log

  # cat /var/spool/cron/crontabs/root
  cron && tail -f /var/log/cron.log
}

# Dump the database to a file and push it to S3
backup() {
  # Dump database to directory
  echo "Backing up $DATABASE to $BACKUP_PATH"
  if [ -d $BACKUP_PATH ]; then
    rm -rf $BACKUP_PATH
  fi
  mkdir -p $BACKUP_PATH
  influxd backup -portable -database $DATABASE -host $DATABASE_HOST:$DATABASE_PORT $BACKUP_PATH
  if [ $? -ne 0 ]; then
    echo "Failed to backup $DATABASE to $BACKUP_PATH"
    exit 1
  fi

  # Compress backup directory
  if [ -e $BACKUP_ARCHIVE_PATH ]; then
    rm -rf $BACKUP_ARCHIVE_PATH
  fi
  tar -cvzf $BACKUP_ARCHIVE_PATH $BACKUP_PATH

  # Push backup file to S3
  echo "Sending file to S3"
  if aws s3 rm s3://${S3_BUCKET}/${S3_KEY_PREFIX}latest.tgz; then
    echo "Removed latest backup from S3"
  else
    echo "No latest backup exists in S3"
  fi
  if aws s3 cp $BACKUP_ARCHIVE_PATH s3://${S3_BUCKET}/${S3_KEY_PREFIX}latest.tgz; then
    echo "Backup file copied to s3://${S3_BUCKET}/${S3_KEY_PREFIX}latest.tgz"
  else
    echo "Backup file failed to upload"
    exit 1
  fi
  if aws s3api copy-object --copy-source ${S3_BUCKET}/${S3_KEY_PREFIX}latest.tgz --key ${S3_KEY_PREFIX}${DATETIME}.tgz --bucket $S3_BUCKET; then
    echo "Backup file copied to s3://${S3_BUCKET}/${S3_KEY_PREFIX}${DATETIME}.tgz"
  else
    echo "Failed to create timestamped backup"
    exit 1
  fi

  echo "Done"
}

# Pull down the latest backup from S3 and restore it to the database
restore() {
  # Remove old backup file
  if [ -d $BACKUP_PATH ]; then
    echo "Removing out of date backup"
    rm -rf $BACKUP_PATH
  fi
  if [ -e $BACKUP_ARCHIVE_PATH ]; then
    echo "Removing out of date backup"
    rm -rf $BACKUP_ARCHIVE_PATH
  fi
  # Get backup file from S3
  echo "Downloading latest backup from S3"
  if aws s3 cp s3://${S3_BUCKET}/${S3_KEY_PREFIX}latest.tgz $BACKUP_ARCHIVE_PATH; then
    echo "Downloaded"
  else
    echo "Failed to download latest backup"
    exit 1
  fi

  # Extract archive
  tar -xvzf $BACKUP_ARCHIVE_PATH -C /

  # Restore database from backup file
  echo "Running restore"
  if influxd restore -database $DATABASE -datadir $DATABASE_DATA_DIR -metadir $DATABASE_META_DIR $BACKUP_PATH ; then
    echo "Successfully restored"
  else
    echo "Restore failed"
    exit 1
  fi
  echo "Done"

}

# Handle command line arguments
case "$1" in
  "startcron")
    startcron "$CRON"
    ;;
  "backup")
    backup
    ;;
  "restore")
    restore
    ;;
  *)
    echo "Invalid command '$@'"
    echo "Usage: $0 {backup|restore|startcron}"
esac


.projenrc.js
const { NodeProject, ProjectType, DockerCompose } = require('projen');

const project = new NodeProject({
  authorAddress: 'damadden88@googlemail.de',
  authorName: 'Martin Mueller',
  name: 'influxdb-s3-backup',
  releaseBranches: 'main',
  projectType: ProjectType.LIB,
});

project.releaseWorkflow.addJobs({
  publish_docker_hub: {
    needs: 'build',
    'runs-on': 'ubuntu-latest',
    env: {
      CI: "true",
    },
    'steps': [
      {
        name: 'Check out the repo',
        uses: 'actions/checkout@v2',
      },
      {
        name: 'Set up QEMU',
        uses: 'docker/setup-qemu-action@v1',
      },
      {
        name: 'Set up Docker Buildx',
        uses: 'docker/setup-buildx-action@v1',
      },
      {
        name: 'Login to DockerHub',
        uses: 'docker/login-action@v1',
        with:{
          username: '${{ secrets.DOCKER_USERNAME }}',
          password: '${{ secrets.DOCKER_PASSWORD }}',
        }
      },
      {
        name: 'get_version',
        id: 'get_version',
        run: [
          'DVERSION=$(jq .version version.json -r)',
          'echo "::set-output name=dversion::$DVERSION"',
          ].join('\n'),
      },
      {
        name: 'Build and push',
        uses: 'docker/build-push-action@v2',
        with: {
          context: '.',
          file: './Dockerfile',
          platforms: 'linux/amd64,linux/arm64',
          push: true,
          tags: 'damadden88/influxdb-s3-backup:${{ steps.get_version.outputs.dversion }},damadden88/influxdb-s3-backup:latest'
        }
      },
    ],
  }
})

new DockerCompose(project, {
  version: '3.3',
  services: {
    influxdb:{
      // container_na me: influxdb,
      image: "influxdb:latest",
      environment: {
        INFLUXDB_DB: 'mydb',
        INFLUXDB_BIND_ADDRESS: '0.0.0.0:8088',
      },
    },
    influxdbs3backup: {
      imageBuild: {
        context: '.',
      },
      environment: {
        DATABASE: 'mydb',
        DATABASE_HOST: 'influxdb',
        S3_BUCKET: 'rasp4backup',
        // AWS_ACCESS_KEY_ID: '',
        // AWS_SECRET_ACCESS_KEY: '',
        AWS_DEFAULT_REGION: 'us-west-2',
      },
    },
  },
});

project.synth();
